include/arch/arm/arch/64/mode/machine.h

```diff
@@ -350,28 +350,61 @@ static inline word_t PURE getFAR(void)
 static inline word_t ats1e2r(word_t va)
 {
     word_t par;
+
+    UNUSED word_t saved_par;
+    if (config_set(CONFIG_ARM_HYPERVISOR_SUPPORT)) {
+        MRS("par_el1", saved_par);
+        isb();
```

Comment:
Is this `isb` necessary? I understand the one before reading PAR, but surely the `at` isn't overtaking the PAR read?

If it is necessary, isn't an `isb` also necessary before writing the saved value back?

Reply:
I dont find anything that backs me up here, but I'm afraid the CPU can execute this in a different order, and consequently change what we want to do.

I added it defensively, I tried to look at the spec, but didn't find anything. I'm okay if we drop it, we just might get exposed to something I'm not sure.

@Indanz I'm open to remove it, if you think it is the best.

Reply:
All I could find in official ARM documentation is that you need an ISB after writing to a system register (because it's a context altering operation). However, AT is special and it does need an ISB before the MSR for obvious reasons, but I didn't find clear documentation for that. Apple code doesn't has an ISB, but someone in the OpenWRT community found out they needed an ISB before reading PAR, or else it didn't work on their hardware.

All in all I agree with you that we should err on the side of caution. That means we should do an ISB between all operations that read or alter PAR. So we're missing another ISB before writing back the saved PAR, as we don't want it to be reordered with the MRS before it.